Detectors/Base/include/DetectorsBase/Track.h

```diff
@@ -63,158 +70,140 @@ namespace AliceO2 {
+          float GetSigmaTglSnp()               const { return mC[kSigTglSnp]; }
+          float GetSigmaTgl2()                 const { return mC[kSigTgl2]; }
+          float GetSigma1PtY()                 const { return mC[kSigQ2PtY]; }
+          float GetSigma1PtZ()                 const { return mC[kSigQ2PtZ]; }
+          float GetSigma1PtSnp()               const { return mC[kSigQ2PtSnp]; }
+          float GetSigma1PtTgl()               const { return mC[kSigQ2PtTgl]; }
+          float GetSigma1Pt2()                 const { return mC[kSigQ2Pt2]; }
+
+          void  Print()                        const;
+
+          // parameters + covmat manipulation
           bool  Rotate(float alpha);
           bool  PropagateTo(float xk, float b);
-          bool  PropagateTo(float xk, const float b[3]);
+          bool  PropagateTo(float xk, const array<float,3> &b);
           void  Invert();
 
-          float GetPredictedChi2(const float p[2], const float cov[3]) const;
-          bool  Update(const float p[2], const float cov[3]);
+          float GetPredictedChi2(const array<float,2> &p, const array<float,3> &cov) const;
+          bool  Update(const array<float,2> &p, const array<float,3> &cov);
 
           bool  CorrectForMaterial(float x2x0,float xrho,float mass,bool anglecorr=false,float dedx=kCalcdEdxAuto);
 
           void  ResetCovariance(float s2=0);
           void  CheckCovariance();
-          void  Print();
 
         protected:
-          // internal cast to TrackPar
-          const TrackPar* Param()              const { return reinterpret_cast<const TrackPar*>(this); }
-          TrackPar* Param()                          { return reinterpret_cast<TrackPar*>(this); }
-          bool TrackPar2Momentum(float p[3], float alpha);
+          float mC[kCovMatSize] = {0.f};  // 15 covariance matrix elements
 
-        protected:
-          float mPC[kTrackPCSize];  // x, alpha + 5 parameters + 15 errors
-
-          static const float kCalcdEdxAuto; // value indicating request for dedx calculation
       };
 
+      class TrackPar final : public TrackParBase { // track parameterization only
+        public:
+          TrackPar() {}
+          TrackPar(float x,float alpha, const array<float,kNParams> &par) : TrackParBase{x,alpha,par} {}
+          TrackPar(const array<float,3> &xyz, const array<float,3> &pxpypz,int sign, bool sectorAlpha=true);
+          //
+          void  Print() const {PrintParam();}
+      };
 
       //____________________________________________________________
-      inline TrackPar::TrackPar(float x, float alpha, const float par[kNParams]) {
+      inline TrackParBase::TrackParBase(float x, float alpha, const array<float, kNParams> &par) : mX{x}, mAlpha{alpha} {
         // explicit constructor
-        mP[kX] = x;
-        mP[kAlpha] = alpha;
-        memcpy(&mP[kY],par,kNParams*sizeof(float));
+        std::copy(par.begin(), par.end(), mP);
```

Comment:
@ktf: for large data (like tracks) the overhead of serialization/deserialization is too large as it is flying through the system - transient in this case means it has to be "messaged" over to many consumers - calibration, QA, whathaveyou. This overhead is both CPU and memory - this is why we don't use any serialization for transient data in the HLT (in extreme cases like ROOT packing, the ESD serialize/deserialize overhead is about the same as full event reconstruction in terms of CPU time). Personally i think this might even work with gcc on x86 but I still don't see any layout guarantees in the standard, so it makes me uncomfortable. Plain arrays make the intent of having a flat buffer more visible in my opinion, so it could be more maintainable; imagine that at some point someone decides to replace std::array with std::vector, might go unnoticed in the review. Let's discuss it in CWG4 if you feel very strongly about having std::array as a store.